src/main/java/sparksoniq/jsoniq/runtime/iterator/RuntimeIterator.java

```diff
@@ -103,9 +114,24 @@ protected RuntimeIterator(List<RuntimeIterator> children, IteratorMetadata metad
         return values;
     }
 
-    protected boolean _hasNext;
-    protected boolean _isOpen;
-    protected List<RuntimeIterator> _children;
-    protected DynamicContext _currentDynamicContext;
-    private final IteratorMetadata metadata;
+    protected List<Item> getItemsFromIteratorWithCurrentContext(RuntimeIterator iterator) {
+        List<Item> result = new ArrayList<>();
+        iterator.open(_currentDynamicContext);
+        while (iterator.hasNext())
+            result.add(iterator.next());
+        iterator.close();
+        return result;
+    }
+
+    protected <T extends Item> T getSingleItemOfTypeFromIterator(RuntimeIterator iterator, Class<T> type) {
```

Comment:
Curios what you think about this. It's not the best generic solution but this is basic functionality that is used in A LOT of iterators that can be simplified

Reply:
Yes, refactoring for functionality used at many places is a great idea.

Regarding functions: in general, we need functionality that checks function signatures in a generic way, not only for single items of a given type. On a high level, there should be somewhere a catalogue of available functions, with their names, arity and signatures (as many sequence types as the arity), and a pointer to the corresponding runtime iterator class.

When a function is called, it is looked up in this catalogue (by name and arity), the corresponding runtime iterator is picked and added to the expression tree, and parameter types are checked at runtime against the signature. If the function isn't found, an error can be through already statically, which makes the code faster at runtime as there is no need to check for the number of parameters any more.

This signature check can be done with fully factored out code, like children iterators that wrap the "actual" children iterators and check the sequence types and may do some numeric casts (s.t. decimal to double, etc).

This kind of architecture can be reused later as well, when we support user-defined functions.